acasclient/selfile.py

```diff
@@ -0,0 +1,847 @@
+        return self._expt_df.copy()
+    @expt_df.setter
+    def expt_df(self, new_expt_df):
+        self._expt_has_changed = self._expt_df.equals(new_expt_df)
+        self._expt_df = new_expt_df
+
+    @property
+    def calculated_results_df(self):
+        """
+        pd.DataFrame container for the 'Calculated Results'.
+        :note:  The df.columns are just the 'Corporate Batch ID' and endpoints, rows represent the values to register.  See calculated_results_datatype for the 'Datatype' mapping of the endpoints to type.
+        """
+        return self._expt_df.copy()
+
+    @calculated_results_df.setter
+    def calculated_results_df(self, new_expt_df):
+        self._expt_has_changed = self._expt_df.equals(new_expt_df)
+        self._expt_df = new_expt_df
+
+
+    @property
+    def datatype(self):
+        """OrderedDict {(column1: datatype1), ...)} column name keys for datatype."""
+        return self._datatype
+    @datatype.setter
+    def datatype(self, new_dict):
+        self._datatype_has_changed = (self._datatype == new_dict)
+        self._datatype = new_dict
+
+
+    @property
+    def calculated_results_datatype(self):
+        """OrderedDict {(column1: datatype1), ...)} column name keys for Calculated Results datatype."""
+        return self._datatype
+    @calculated_results_datatype.setter
+    def calculated_results_datatype(self, new_dict):
+        self._datatype_has_changed = (self._datatype == new_dict)
+        self._datatype = new_dict
+
+
+class Generic(AbstractExperiment):
+    """
+    Class to represent a Generic ACAS experiment file.
+    Logical units (meta data, datatype, and calculated_results) are
+    represented as separate data members.  There is some internal
+    tracking to detect if the original data content is altered, but
+    the datatype changes need to be updated by the caller.
+    calculated_results_datatype - Datatype endpoint mapping.
+    calculated_results_df - Calculated Results dataframe.
+    Common formats are supported for both input and output:  csv,
+    and excel
+    API examples:
+    # Read an existing file, change the 'Project' metadata and save.
+    import datateam.fileio.acas as acas
+    expt = acas.Generic('35250_FG_acas.csv')
+    expt.project = 'Test Project'
+    expt.saveAs('35250_FG_modified.xlsx')
+    # Change the experimental values and save.
+    import pandas as pd
+    new_results = pd.read_csv('new_data.csv')
+    expt.calculated_results_df = new_results
+    expt.saveAs('replaced_data.csv')
+    # Create new, completely empty handle, and populate it.
+    expt = acas.Generic()
+    expt.protocol_name = 'my super cool protocol'
+    expt.experiment_name = 'ATP-Glo Assay 2020-05-19'
+    expt.scientist = 'Jackson'
+    expt.notebook = 'results_2020-05-19.pdf'
+    expt.page = 'page 19'
+    expt.assay_date = '2020-05-20'
+    expt.project = 'Star89'
+    # Dummy data, but this could come from a ctfile.sd_to_df, or
```

Comment:
I think the API examples are good to keep, and verify.  But I realize that this ctfile module references something that is only our data team library.  Not critical, but removing the reference to ctfile might reduces questions from users down the line.  You may also want to consider lifting that code too, someday.  Since any other (non-text/clob) SD reader is almost certainly lossy in some, subtle, hard to debug, way.

Reply:
Thanks, make sense to look at lifting that code too in the future.  I modified this line slightly to remove the reference to ctfile.sd_to_df